blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +369,256 @@ where
+        let cid = message
+            .cid()
+            .map_err(|e| Error::Other(format!("Could not get cid from message {:?}", e)))?;
+
+        let cid_for_task = cid.clone();
+        let address_for_task = *message.from();
+        let sequence_for_task = message.sequence();
+        let height_of_head = tipset.epoch();
+        let task = task::spawn(async move {
+            let (back_t, back_r) = Self::search_back_for_message(
+                block_store_for_message,
+                &tipset,
+                (&address_for_task, &cid_for_task, &sequence_for_task),
+            )?
+            .ok_or_else(|| {
+                Error::Other("State manager not subscribed to back search wait".to_string())
+            })?;
+            let back_tuple = (back_t, back_r);
+            sender
+                .send(())
+                .map_err(|e| Error::Other(format!("Could not send to channel {:?}", e)))?;
+            Ok::<_, Error>(back_tuple)
+        });
+
+        let mut reverts: HashMap<TipsetKeys, bool> = HashMap::new();
+        while let Some(subscriber) = subscriber
+            .clone()
+            .ok_or_else(|| {
+                Error::Other("State Manager not subscribed to tipset head changes".to_string())
+            })?
+            .next()
+            .await
+        {
+            match subscriber {
+                HeadChange::Revert(_tipset) => {
+                    if candidate_tipset.is_some() {
+                        candidate_tipset = None;
+                        candidate_receipt = None;
+                    }
+                }
+                HeadChange::Apply(tipset) => {
+                    if candidate_tipset
+                        .as_ref()
+                        .map(|s| s.epoch() >= s.epoch() + tipset.epoch())
+                        .unwrap_or_default()
+                    {
+                        return Ok((candidate_tipset, candidate_receipt));
+                    }
+                    let poll_receiver = receiver.try_recv().map_err(|e| {
+                        Error::Other(format!("Could not receieve from channel {:?}", e))
+                    })?;
+                    if poll_receiver.is_some() {
+                        reverts.insert(tipset.key().to_owned(), true);
+                    }
+
+                    let message_var = (message.from(), &message.sequence());
+                    let maybe_receipt =
+                        Self::tipset_executed_message(&*block_store, &tipset, &cid, message_var)?;
+                    if let Some(receipt) = maybe_receipt {
+                        if confidence == 0 {
+                            return Ok((Some(tipset), Some(receipt)));
+                        }
+                        candidate_tipset = Some(tipset);
+                        candidate_receipt = Some(receipt)
+                    }
+                }
+                _ => (),
+            }
+        }
+        let (back_tipset, back_receipt) = task.await?;
```

Comment:
the only thing is this should be awaiting on along with the subscriber channel. The point of this is that whichever comes first should be polled, so they should be polled at the same time, instead of waiting for one to complete then awaiting on this